Image Banner updates

LucasLacerdaUX opened the pull request:
Work in Progress

**Why are these changes introduced?**
Improves the Image Banner section with new settings & customizability options.

Closes #216 

**What approach did you take?**
This PR adds the following settings:
_Image banner section_
- `Show text box in desktop`
- `Image overlay opacity`
- Added `Visible when text box displayed` info text to the `Color Scheme` setting
- Added `Mobile layout` header
- `Show text below images`

_Heading Block_
- Added `Heading size`

**Other considerations**

**Demo links**
- [Store](https://os2-demo.myshopify.com/?preview_theme_id=124616376342)
- [Editor](https://os2-demo.myshopify.com/admin/themes/124616376342/editor)

**Checklist**
- [x] Followed [theme code principles](https://github.com/Shopify/dawn/blob/main/.github/CONTRIBUTING.md#theme-code-principles)
- [x] Linted with [Theme Check](https://github.com/Shopify/theme-check)
- [x] Tested on [mobile](https://shopify.dev/themes/store/requirements#mobile-browser-requirements)
- [x] Tested on [multiple browsers](https://shopify.dev/themes/store/requirements#desktop-browser-requirements)
- [x] Tested for [accessibility](https://shopify.dev/themes/best-practices/accessibility)

LucasLacerdaUX commented:
@Oliviammarcello this should now be ready for design review. Let me know if it's working as expected. Not every scenario was covered in the original designs so maybe I missed something. If you prefer to pair to make adjustments, let me know :)

melissaperreault commented on locales/en.default.schema.json:
```diff
@@ -675,11 +681,18 @@
           },
-          "label": "Color scheme"
+          "label": "Color scheme",
+          "info": "Visible when text box displayed"
```
@gretchen-willenberg The image banner updates are not merged yet, but I don't see the text `Hidden when text box displayed` as raised in your audit, is it possible that this has changed already to another form?

cc. @Oliviammarcello

Oliviammarcello commented:
@LucasLacerdaUX this looks amazing!! :) 

I just have a few notes:
- [ ] When "Adapt section height to first image size" is selected and narrow images are used, the content can push past the images. [Shown here.](https://screenshot.click/21-17-zocu3-xwsq1.mp4) Is there a way to ensure that the images stretch to fit the content so this doesn't happen? 
- [ ] When "Adapt section height to first image size" is selected and on mobile, the text is below the images, there is an awkward white space that shouldn't be there. [Video.](https://screenshot.click/21-45-1uroh-jdkrg.mp4)
- [ ] When there are two images stacked on mobile without the text box, the text should be aligned to the center of the first image. [Screenshot.](https://screenshot.click/21-42-r705u-kjd6j.png)
- [ ] On tablet, the box goes below the image to match what is done on mobile. The text box should remain on top (like on desktop) until the mobile viewport. Sorry if the Figma file was misleading. [Figma link.](https://www.figma.com/file/JcYdswrk2JhqBTGBVWPM7J/Dawn-2.0?node-id=2518%3A51810)

Let me know if you need to pair or have any questions! Again, this is looking really great!! :)

LucasLacerdaUX commented:
I think I managed to fix the main issues you reported above, @Oliviammarcello, except for:

> On tablet, the box goes below the image to match what is done on mobile. The text box should remain on top (like on desktop) until the mobile viewport. Sorry if the Figma file was misleading. Figma link.

I didn't get what you mean by that. Right now, the tablet and desktop layout will look the same. In Figma, they look the same to me 👀 

![image](https://user-images.githubusercontent.com/37168033/129808864-1adf6111-250b-4532-8bee-7d0d3b3fda7e.png)

Can you share some Actual vs Expected screenshots or a setting combination that produces a wrong layout just so I know what's going on?

> When "Adapt section height to first image size" is selected and narrow images are used, the content can push past the images. Shown here. Is there a way to ensure that the images stretch to fit the content so this doesn't happen?

With pure CSS, there isn't... I managed to fix the content pushing past this area, but that created another issue:
![image](https://user-images.githubusercontent.com/37168033/129809721-0a5c069d-b2a8-4185-a28e-2073f1001f10.png)

I'll try to find a fix for that tomorrow. Let me know if you find any new issues.

Oliviammarcello commented:
> I think I managed to fix the main issues you reported above, @Oliviammarcello, except for:
> 
> > On tablet, the box goes below the image to match what is done on mobile. The text box should remain on top (like on desktop) until the mobile viewport. Sorry if the Figma file was misleading. Figma link.
> 
> I didn't get what you mean by that. Right now, the tablet and desktop layout will look the same. In Figma, they look the same to me 👀

On the live version, the tablet screens match what the mobile screens look like, not what the desktop looks like. [Video.](https://screenshot.click/21-01-3j99n-h6rca.mp4) The screenshot you added is correct. 

> Can you share some Actual vs Expected screenshots or a setting combination that produces a wrong layout just so I know what's going on?
> 
> > When "Adapt section height to first image size" is selected and narrow images are used, the content can push past the images. Shown here. Is there a way to ensure that the images stretch to fit the content so this doesn't happen?
> 
Force image to adapt to content. I think we do this in the image with text section as well. 
![image](https://user-images.githubusercontent.com/60230008/129903802-eec08efd-a025-4ff1-b481-0f70650ed8e7.png)

Oliviammarcello commented:
- [ ] The padding above the text is missing on mobile.[ Screenshot.](https://screenshot.click/21-12-tlyhw-lc1nv.png) | [Figma](https://www.figma.com/file/JcYdswrk2JhqBTGBVWPM7J/Dawn-2.0?node-id=4828%3A73282)

Oliviammarcello commented:
- [ ] The section hides the content below it when images are stacked and content is below. [Video.](https://screenshot.click/21-18-tcrje-zmos0.mp4)

svinkle commented:
Can we consider adding a setting to include a drop-shadow for text without a solid background? The readability in the example above is quite poor. A drop-shadow would help provide contrast between the text and the background image.

Oliviammarcello commented:
> Can we consider adding a setting to include a drop-shadow for text without a solid background? The readability in the example above is quite poor. A drop-shadow would help provide contrast between the text and the background image.

There is an image opacity slider that merchants can use to increase contrast @svinkle

LucasLacerdaUX commented on sections/image-banner.liquid:
```diff
@@ -57,12 +70,12 @@
       >
```
Only when `Show text below images on mobile` is enabled:

![image](https://user-images.githubusercontent.com/37168033/130466551-78840533-48ec-4310-b166-dc0d312119b7.png)

LucasLacerdaUX commented on sections/image-banner.liquid:
```diff
@@ -7,10 +7,23 @@
+    opacity: {{ section.settings.image_overlay_opacity | divided_by: 100.0 }};
+  }
+{%- endstyle -%}
+
```
That should be the expected layout (matching the height of the first element), but there's something wrong with the buttons positioning. I'll double check the implementation for that.

LucasLacerdaUX commented on assets/section-image-banner.css:
```diff
@@ -72,7 +70,46 @@
+  .banner--stacked:not(.banner--mobile-bottom) .banner__media {
+    position: relative;
+  }
+
```
Yeah, that's a side-effect of matching the container height to the first image height. Do you have any suggestions on how to fix that? 👀 

I honestly have no idea .-.

sofiamatulis replied:
I noticed this is also happening on desktop: https://screenshot.click/23-01-nsavn-03qb2.png
 I think we can just add the 39rem height to just media query of mobile: https://screenshot.click/23-02-0uxuj-2vltf.mp4

As for mobile, this is only an issue when you add the 39rem height when there are two images: https://screenshot.click/23-07-n1z2c-lxpio.mp4

I think you may want to remove it when there are two? 🤔

LucasLacerdaUX replied:
Thank you so much for this detailed comment, Thomas. It helped a lot. 

I decided to rewrite my approach a little bit and add the `padding-top` to `content` on `mobile` (absolute positioning the `media`), and to `media` on `desktop` (absolute positioning the media).

It's MUCH better now, as the content will scale most of the times, except for:
- <img src="https://user-images.githubusercontent.com/37168033/130546419-db956e9a-80f3-40e2-9506-2d164f5604c2.png" width="250">
- <img src="https://user-images.githubusercontent.com/37168033/130546406-4ec87694-2550-4820-9008-7c1204141b1e.png" width="200">
- Long content inside `Heading Block`
- Wide images

But for that specific scenario, I still couldn't figure a solution, as the images CAN'T be absolute positioned (otherwise, it wouldn't follow `Adapt section height to first image size`) and therefore, the content needs to be `absolute` positioned to stay above the image.

That's the best I could do for now and I can't find a solution using CSS + Liquid. Not sure if this is considered an "edge case" or acceptable.

Oliviammarcello replied:
To add context to the decision to not use an RTE here; we intentionally decided to allow merchants to add a lot of text to image with text and not to image banner. Since the content covers the image on the banner, we don't want to encourage merchants to add anything more than a few words. Because of this decision, if merchants wanted an opportunity to write more text and want it to be paired with an image they have the image with text section. The text block is disabled by default in the banner section for this reason as well.

LucasLacerdaUX replied:
@sofiamatulis can you please share the combination of settings that led you to these screenshots? I couldn't reproduce that on my end. Here's what it looks like on mobile with:

![image](https://user-images.githubusercontent.com/37168033/130718137-d38bf051-fb64-4046-95c7-9fa72e9e3e89.png)


<img src="https://user-images.githubusercontent.com/37168033/130718082-53ed74f5-9dcc-4929-a22a-7f0b23b524dd.png" width="270">
<img src="https://user-images.githubusercontent.com/37168033/130718160-58476729-c5bc-4b11-81be-5de1a93a9c2d.png" width="270">

melissaperreault commented:
> We will need to check with Melissa or Olivia to look at it because this results having no equal width anymore with 2 buttons on mobile

I am not sure I understand the main concern, is it because when button wraps [currently they are misaligned](https://screenshot.click/21-08-mewzp-pfh51.png)? 

When buttons are side by side, they can have various length. It's when they wrap that we ideally wanted them to grow together the same width. I recall discussing this with @sofiamatulis I believe, but can't remember more! 😅 

Unless there are no other path to maintain the same width when they wrap, I would be ok to have different width. Thoughts @Oliviammarcello ?

melissaperreault commented:
> The solution I had in mind is adding an arbitrary height value that is reasonable for 95% of merchants (unlike us where we test extreme cases)
Maybe could we talk with Melissa/Olivia about this?

Ready to chat/huddle to understand better the concern if it can unblock! Do not hesitate to ping one of us on slack! cc. @Oliviammarcello

LucasLacerdaUX commented on assets/section-image-banner.css:
```diff
@@ -2,7 +2,10 @@
+.banner--mobile-bottom:not(.banner--stacked):not(.banner--adapt) {
+  height: 39rem;
 }
 
```
Thank you so much Thomas! <3

I've added the new properties with `gap`, removed the margin like you mentioned, but kept `flex-grow`. It seems to both fix the issue AND allow for the button to auto-expand when they're in separate lines.

Let me know if that works.

LucasLacerdaUX commented:
I've made some updates but I couldn't fix all the bugs.

I did my best to fix all the issues, but I simply don't know how to do that anymore. Every time I fix a bug, it breaks the layout in another combination of settings ><'

I have just pushed some updates to this branch. It fixed pretty much every bug that has been reported, except for [this set of settings](https://github.com/Shopify/dawn/pull/384#discussion_r694445176). I'd love some help tomorrow on fixing that.

tauthomas01 commented on assets/section-image-banner.css:
```diff
@@ -118,10 +168,42 @@
-  margin: 0 auto;
-  margin-top: 2.3rem;
-  transform: translateX(1rem);
+  align-items: baseline;
```
I think we forgot to add `display: flex` because the gap isn't working

```suggestion
  align-items: baseline;
  display: flex;
```

![image](https://user-images.githubusercontent.com/658169/130630164-d4e743a9-bb1f-4431-97ad-7b116f5676ea.png)

tauthomas01 reviewed (commented):
Works great!

> I just pushed some updates to this branch that fixed pretty much every bug that has been reported so far, EXCEPT for this set of settings.

I am unable to reproduce this edge case: it seems to scale well even if I add a lot of text. Here's my setup:

![image](https://user-images.githubusercontent.com/658169/130640022-cd770bd1-1642-4296-9bf0-b5803a84da2b.png)

sofiamatulis commented on assets/section-image-banner.css:
```diff
@@ -2,10 +2,13 @@
   display: flex;
```
In the default state there isnt space between the two buttons: https://screenshot.click/24-46-wkm79-tduo9.png

LucasLacerdaUX replied:
My bad. I missed the `display: flex` on the container. Just updated that.

LucasLacerdaUX commented:
@tauthomas01 @sofiamatulis Just did another update. PR should now be ready for a final review.

- Updated the buttons to use `display: flex` so `gap` works.
- Fixed a bug (https://github.com/Shopify/dawn/pull/384#discussion_r695350927) where the content was being cut from the container, despite having enough space to fit in.

The only remaining issue (https://github.com/Shopify/dawn/pull/384#discussion_r695350927) I now consider to be an edge case + side effect of the merchant choosing the wrong settings. We are following every setting in our logic, it's just that the merchant served a big amount of content to fit in a narrow image.

Let me know if there are still any outstanding issues I missed and if you're okay.

P.S.: This section is officially a developer's nightmare. I feel like I need to do a CSS course after working on this so hopefully I can have some ideas for a refactor later 💀

sofiamatulis reviewed (dismissed):
All scenarios look good and work well 🎉 I would consider the scenarios we talked about to be

1- edge cases
2- actually working as expected, so the behaviour is correct and it's more up to the merchant to select the right image and setting and not to add too much texr

tauthomas01 reviewed (dismissed):
Tested on Firefox, Chrome and Safari + desktop/mobile. Everything looks good.
Hopefully I didn't miss something

Oliviammarcello commented on assets/section-image-banner.css:
```diff
@@ -98,14 +135,34 @@
   display: flex;
```
I agree the flex should only be there when there are two buttons on mobile

Oliviammarcello replied:
It shouldn't be there on desktop

tauthomas01 commented on assets/section-image-banner.css:
```diff
@@ -98,14 +135,34 @@
 @media screen and (min-width: 750px) {
   .banner__content {
     padding-bottom: 5rem;
     padding-top: 5rem;
```
Now I see why I couldn't reproduce it... I never thought of using a logo as a background image.

LucasLacerdaUX replied:
Yeah, I spent so much time on this one, but in the end, I feel like the section is working exactly as expected. If we force a taller height, we're no longer `adapting to the first image`. 

So I decided to simply add an `overflow: hidden` so the content won't be placed above other sections of the page, in case it overflows.

LucasLacerdaUX commented on sections/image-banner.liquid:
```diff
@@ -57,12 +76,12 @@
       >
     </div>
   {%- endif -%}
-  <div class="banner__content page-width" style="align-items: {{ section.settings.desktop_text_box_position }};">
```
Oh, okay. Just remembered why I made this change.

It's because it's now applied on Desktop only, under a `@media screen and (min-width: 750px)` query. On mobile, we'll always use `align-items: center`.

sofiamatulis replied:
That makes sense 🙏 thanks for the explanation

sofiamatulis reviewed (approved):
🚢

Oliviammarcello commented:
@PaulNewton 

Our intention with this section is to provide a lot of flexibility to adapt to the type of assets merchants have. The settings `Show text box on desktop` and `Show text below images` on mobile are enabled by default to ensure accessibility. The opacity slider is a fallback for merchants who want to remove the box, but need to darken their images slightly. We stand on the side of giving merchants the tools they need to be successful and trust they’ll make good choices.
